In [26]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns
sns.set_style('whitegrid')
%matplotlib inline 

In [27]:
train = pd.read_csv('../input/train.csv')
test = pd.read_csv('../input/test.csv')

In [28]:
print(train.shape)
print(test.shape)

(11769, 15)
(5789, 13)


In [29]:
train.head()

no                                모델명           연월      연식   연료   주행거리  \
0   0        기아 더 뉴 K7 2.4 GDI 프레스티지 스페셜  12/12(13년형)  2013.0  가솔린   4만km   
1   1          현대 YF쏘나타 2.0 Y20 LPi 프리미어        10/03  2010.0  LPG   9만km   
2   2               현대 그랜져HG 220 디젤 프리미엄  14/09(15년형)  2015.0   디젤   3만km   
3   3            쌍용 뉴카이런 2.0 LV5 2WD 고급형  07/06(08년형)  2008.0   디젤  16만km   
4   4  현대 뉴스타렉스 점보 밴 TCI 3인승 GX 윈도우밴 일반형        04/03  2004.0   디젤  16만km   

    인승  최대출력(마력)   기통  최대토크(kgm) 구동방식 자동수동 국산/수입  신차가(만원)  가격(만원)  
0  NaN     201.0  4.0       25.5   FF  NaN    국산   3141.0  1870.0  
1  NaN     157.0  NaN       20.0   FF  NaN    국산      NaN   700.0  
2  NaN     202.0  4.0       45.0   FF  NaN    국산   3389.0  2990.0  
3  NaN     151.0  4.0       33.8   FR  NaN    국산   2499.0   420.0  
4  3.0     103.0  NaN       24.0   FR   수동    국산   1445.0   450.0

In [30]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11769 entries, 0 to 11768
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   no         11769 non-null  int64  
 1   모델명        11769 non-null  object 
 2   연월         11769 non-null  object 
 3   연식         11769 non-null  float64
 4   연료         11769 non-null  object 
 5   주행거리       11769 non-null  object 
 6   인승         765 non-null    float64
 7   최대출력(마력)   10644 non-null  float64
 8   기통         8698 non-null   float64
 9   최대토크(kgm)  10545 non-null  float64
 10  구동방식       10808 non-null  object 
 11  자동수동       1313 non-null   object 
 12  국산/수입      11769 non-null  object 
 13  신차가(만원)    9544 non-null   float64
 14  가격(만원)     11769 non-null  float64
dtypes: float64(7), int64(1), object(7)
memory usage: 1.3+ MB


In [31]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5789 entries, 0 to 5788
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   no         5789 non-null   int64  
 1   모델명        5789 non-null   object 
 2   연월         5789 non-null   object 
 3   연식         5789 non-null   float64
 4   연료         5789 non-null   object 
 5   주행거리       5789 non-null   object 
 6   인승         389 non-null    float64
 7   최대출력(마력)   5237 non-null   float64
 8   기통         4275 non-null   float64
 9   최대토크(kgm)  5183 non-null   float64
 10  구동방식       5312 non-null   object 
 11  자동수동       665 non-null    object 
 12  국산/수입      5789 non-null   object 
dtypes: float64(5), int64(1), object(7)
memory usage: 588.1+ KB


# Feature name 영어로 모두 바꿔야 한다

In [32]:
train.columns = ['no', 'modelname', 'ym', 'yeartype', 'fuel', 'dist', 'people', 'power', 'cylinder', 'torque', 'type', 'mission', 'country', 'newprice', 'price']
test.columns = ['no', 'modelname', 'ym', 'yeartype', 'fuel', 'dist', 'people', 'power', 'cylinder', 'torque', 'type', 'mission', 'country']

In [33]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11769 entries, 0 to 11768
Data columns (total 15 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   no         11769 non-null  int64  
 1   modelname  11769 non-null  object 
 2   ym         11769 non-null  object 
 3   yeartype   11769 non-null  float64
 4   fuel       11769 non-null  object 
 5   dist       11769 non-null  object 
 6   people     765 non-null    float64
 7   power      10644 non-null  float64
 8   cylinder   8698 non-null   float64
 9   torque     10545 non-null  float64
 10  type       10808 non-null  object 
 11  mission    1313 non-null   object 
 12  country    11769 non-null  object 
 13  newprice   9544 non-null   float64
 14  price      11769 non-null  float64
dtypes: float64(7), int64(1), object(7)
memory usage: 1.3+ MB


In [34]:
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5789 entries, 0 to 5788
Data columns (total 13 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   no         5789 non-null   int64  
 1   modelname  5789 non-null   object 
 2   ym         5789 non-null   object 
 3   yeartype   5789 non-null   float64
 4   fuel       5789 non-null   object 
 5   dist       5789 non-null   object 
 6   people     389 non-null    float64
 7   power      5237 non-null   float64
 8   cylinder   4275 non-null   float64
 9   torque     5183 non-null   float64
 10  type       5312 non-null   object 
 11  mission    665 non-null    object 
 12  country    5789 non-null   object 
dtypes: float64(5), int64(1), object(7)
memory usage: 588.1+ KB


In [35]:
# sns.heatmap(train.isnull(), cbar=False)

In [36]:
# sns.heatmap(test.isnull(), cbar=False)

In [37]:
train_no = train['no']
test_no = test['no']
train = train.drop(['no'], axis=1)
test = test.drop(['no'], axis=1)

# 이제 하나씩 처리하자
```
0   no         11769 non-null  int64    
1   모델명        11769 non-null  object   
2   연월         11769 non-null  object   
3   연식         11769 non-null  float64  
4   연료         11769 non-null  object   
5   주행거리       11769 non-null  object   
6   인승         765 non-null    float64  
7   최대출력(마력)   10644 non-null  float64  
8   기통         8698 non-null   float64  
9   최대토크(kgm)  10545 non-null  float64  
10  구동방식       10808 non-null  object   
11  자동수동       1313 non-null   object   
12  국산/수입      11769 non-null  object   
13  신차가(만원)    9544 non-null   float64  
14  가격(만원)     11769 non-null  float64  
```

```
0   no         11769 non-null  int64  
1   modelname  11769 non-null  object 
2   ym         11769 non-null  object 
3   yeartype   11769 non-null  float64
4   fuel       11769 non-null  object 
5   dist       11769 non-null  object 
6   people     765 non-null    float64
7   power      10644 non-null  float64
8   cylinder   8698 non-null   float64
9   torque     10545 non-null  float64
10  type       10808 non-null  object 
11  mission    1313 non-null   object 
12  country    11769 non-null  object 
13  newprice   9544 non-null   float64
14  price      11769 non-null  float64
```

# 1 모델명 modelname 11769 non-null object

- 모델명을 그냥 날려보자 이번엔

In [38]:
train = train.drop(['modelname'], axis=1)
test = test.drop(['modelname'], axis=1)

print(train.shape)
print(test.shape)

(11769, 13)
(5789, 11)


# 2 연월 ym 11769 non-null object

- 괄호가 있는 게 있고 없는 게 있다
- 괄호 안의 정보는 필요 없어보인다 (그냥 몇년형이냐 이건데 차피 다 나와있음)
- 괄호를 제거하고
- 음... 이건 범주형인지 연속형인지 모르겠다
- 일단 연/월 나눠서, 각각 칼럼을 연속형이라고 가정하고 넘버로 바꿔서 넣어보자
- 나중에 잘 안 나오면 원핫인코딩 돌려서 한번 더 해보자
- 근데 int64가 맞나? float로 해야 하는 거 아닐까?

In [39]:
train_ym = train['ym']
test_ym = test['ym']

train_ym.head()

0    12/12(13년형)
1          10/03
2    14/09(15년형)
3    07/06(08년형)
4          04/03
Name: ym, dtype: object

In [40]:
train_ym = train_ym.apply(lambda x: x.split('(')[0])
test_ym = test_ym.apply(lambda x: x.split('(')[0])

train_ym.head()

0    12/12
1    10/03
2    14/09
3    07/06
4    04/03
Name: ym, dtype: object

In [41]:
train_year = train_ym.apply(lambda x: float(x.split('/')[0]))
train_month = train_ym.apply(lambda x: float(x.split('/')[1]))

test_year = test_ym.apply(lambda x: float(x.split('/')[0]))
test_month = test_ym.apply(lambda x: float(x.split('/')[1]))

train_year.head()

0    12.0
1    10.0
2    14.0
3     7.0
4     4.0
Name: ym, dtype: float64

In [42]:
train_year.value_counts()

10.0    1140
11.0    1088
9.0     1062
14.0    1047
13.0    1018
12.0    1011
15.0     986
8.0      811
16.0     761
7.0      543
17.0     451
6.0      425
18.0     318
5.0      316
4.0      179
19.0     147
2.0      112
3.0      110
1.0       50
20.0      35
0.0       25
97.0      21
99.0      16
96.0      14
92.0      13
94.0      12
93.0      10
95.0      10
88.0       7
91.0       7
98.0       7
87.0       6
89.0       4
90.0       2
83.0       1
72.0       1
78.0       1
85.0       1
79.0       1
Name: ym, dtype: int64

## year 전처리는 필요하다
- 99에서 00 넘어가는 그 시점
- 30보다 크면 19 붙이고, 아니면 20 붙여서 4자리 만들자

In [43]:
train_year = train_year.apply(lambda x: x + (1900 if x > 30 else 2000))
train_year.head()

0    2012.0
1    2010.0
2    2014.0
3    2007.0
4    2004.0
Name: ym, dtype: float64

In [44]:
train_year.value_counts()

2010.0    1140
2011.0    1088
2009.0    1062
2014.0    1047
2013.0    1018
2012.0    1011
2015.0     986
2008.0     811
2016.0     761
2007.0     543
2017.0     451
2006.0     425
2018.0     318
2005.0     316
2004.0     179
2019.0     147
2002.0     112
2003.0     110
2001.0      50
2020.0      35
2000.0      25
1997.0      21
1999.0      16
1996.0      14
1992.0      13
1994.0      12
1995.0      10
1993.0      10
1998.0       7
1991.0       7
1988.0       7
1987.0       6
1989.0       4
1990.0       2
1978.0       1
1972.0       1
1979.0       1
1985.0       1
1983.0       1
Name: ym, dtype: int64

In [45]:
train_year.name = 'year'
train_month.name = 'month'
test_year.name = 'year'
test_month.name = 'month'

train = train.drop(['ym'], axis=1)
test = test.drop(['ym'], axis=1)

train = train.join([train_year, train_month])
test = test.join([test_year, test_month])

In [46]:
train.head()

yeartype fuel   dist  people  power  cylinder  torque type mission country  \
0    2013.0  가솔린   4만km     NaN  201.0       4.0    25.5   FF     NaN      국산   
1    2010.0  LPG   9만km     NaN  157.0       NaN    20.0   FF     NaN      국산   
2    2015.0   디젤   3만km     NaN  202.0       4.0    45.0   FF     NaN      국산   
3    2008.0   디젤  16만km     NaN  151.0       4.0    33.8   FR     NaN      국산   
4    2004.0   디젤  16만km     3.0  103.0       NaN    24.0   FR      수동      국산   

   newprice   price    year  month  
0    3141.0  1870.0  2012.0   12.0  
1       NaN   700.0  2010.0    3.0  
2    3389.0  2990.0  2014.0    9.0  
3    2499.0   420.0  2007.0    6.0  
4    1445.0   450.0  2004.0    3.0

In [47]:
print(train.shape)
print(test.shape)

(11769, 14)
(5789, 12)


# 3 연식 yeartype 11769 non-null float64

건들지 말자

# 4 연료 fuel 11769 non-null object

- 일단 범주형
- 일단 다 더미로 바꾸고
- 나중에 합쳐진 애들은 하이브리드로 묶어볼까?

In [48]:
train_fuel = train['fuel']
test_fuel = test['fuel']

train_fuel.value_counts()

가솔린          6671
디젤           3967
LPG           934
가솔린+전기         79
가솔린/LPG겸용      41
기타             32
전기             24
LPG+전기         14
가솔린/CNG겸용       4
CNG             2
수소              1
Name: fuel, dtype: int64

In [49]:
test_fuel.value_counts()

가솔린          3280
디젤           1932
LPG           469
가솔린+전기         34
가솔린/LPG겸용      33
전기             19
기타             11
LPG+전기          8
CNG             2
가솔린/CNG겸용       1
Name: fuel, dtype: int64

In [50]:
train_fuel_dummies = pd.get_dummies(train_fuel)
test_fuel_dummies = pd.get_dummies(test_fuel)


# MISSING COLUMN PROCESSING
train_missing_cols = set(test_fuel_dummies.columns) - set(train_fuel_dummies.columns)
for c in train_missing_cols:
    train_fuel_dummies[c] = 0
train_fuel_dummies = train_fuel_dummies[test_fuel_dummies.columns]
test_missing_cols = set(train_fuel_dummies.columns) - set(test_fuel_dummies.columns)
for c in test_missing_cols:
    test_fuel_dummies[c] = 0
test_fuel_dummies = test_fuel_dummies[train_fuel_dummies.columns]
# MISSING COLUMN PROCESSING


train_fuel_dummies = train_fuel_dummies.add_prefix('fuel_')
test_fuel_dummies = test_fuel_dummies.add_prefix('fuel_')

train = train.drop(['fuel'], axis=1)
test = test.drop(['fuel'], axis=1)

train = train.join(train_fuel_dummies)
test = test.join(test_fuel_dummies)

In [51]:
train.head()




yeartype   dist  people  power  cylinder  torque type mission country  \
0    2013.0   4만km     NaN  201.0       4.0    25.5   FF     NaN      국산   
1    2010.0   9만km     NaN  157.0       NaN    20.0   FF     NaN      국산   
2    2015.0   3만km     NaN  202.0       4.0    45.0   FF     NaN      국산   
3    2008.0  16만km     NaN  151.0       4.0    33.8   FR     NaN      국산   
4    2004.0  16만km     3.0  103.0       NaN    24.0   FR      수동      국산   

   newprice  ...  fuel_CNG  fuel_LPG  fuel_LPG+전기  fuel_가솔린  fuel_가솔린+전기  \
0    3141.0  ...         0         0            0         1            0   
1       NaN  ...         0         1            0         0            0   
2    3389.0  ...         0         0            0         0            0   
3    2499.0  ...         0         0            0         0            0   
4    1445.0  ...         0         0            0         0            0   

   fuel_가솔린/CNG겸용  fuel_가솔린/LPG겸용  fuel_기타  fuel_디젤  fuel_전기  
0               0               0        0        0        0  
1               0               0        0        0        0  
2               0               0        0        1        0  
3               0               0        0        1        0  
4               0               0        0        1        0  

[5 rows x 23 columns]

In [52]:
print(train.shape)
print(test.shape)

(11769, 23)
(5789, 21)


# 5 주행거리 dist 11769 non-null object

- %d만{km|ml}
- 중간중간에 '등록'이라고 되어있는 게 있다
- 평균으로 채워야 한다
- 평균을 구하려면 일단 수치를 모두 구해야 한다
- 등록은 스킵해야 하네 

In [53]:
train_dist = train['dist']
test_dist = test['dist']

train_dist.head()

0     4만km
1     9만km
2     3만km
3    16만km
4    16만km
Name: dist, dtype: object

In [54]:
def preprocess_dist(x):
    # '등록'은 나중에 평균으로 채울 거임
    if x == '등록':
        return None

    if x[-2:] == 'km':
        # kilometer
        x = x.split('km')[0]
        weight = 1
    else:
        # mile
        assert x[-2:] == 'ml'
        x = x.split('ml')[0]
        weight = 1.609344

    if x[-1] == '만':
        x = float(x.split('만')[0]) * 10000
    elif x[-1] == '천':
        x = float(x.split('천')[0]) * 1000

    return float(x * weight)


In [55]:
train_dist = train_dist.apply(preprocess_dist)
test_dist = test_dist.apply(preprocess_dist)

train_dist

0         40000.0
1         90000.0
2         30000.0
3        160000.0
4        160000.0
           ...   
11764    110000.0
11765      5000.0
11766       230.0
11767      9000.0
11768      3000.0
Name: dist, Length: 11769, dtype: float64

In [56]:
# 등록 처리
train_dist.isna().sum()

29

In [57]:
train_dist = train_dist.fillna(train_dist.mean())
test_dist = test_dist.fillna(test_dist.mean())

In [58]:
train_dist.isna().sum()

0

In [59]:
train = train.drop(['dist'], axis=1)
test = test.drop(['dist'], axis=1)

train_dist.name = 'dist'
test_dist.name = 'dist'

train = train.join(train_dist)
test = test.join(test_dist)

In [60]:
train.head()

yeartype  people  power  cylinder  torque type mission country  newprice  \
0    2013.0     NaN  201.0       4.0    25.5   FF     NaN      국산    3141.0   
1    2010.0     NaN  157.0       NaN    20.0   FF     NaN      국산       NaN   
2    2015.0     NaN  202.0       4.0    45.0   FF     NaN      국산    3389.0   
3    2008.0     NaN  151.0       4.0    33.8   FR     NaN      국산    2499.0   
4    2004.0     3.0  103.0       NaN    24.0   FR      수동      국산    1445.0   

    price  ...  fuel_LPG  fuel_LPG+전기  fuel_가솔린  fuel_가솔린+전기  fuel_가솔린/CNG겸용  \
0  1870.0  ...         0            0         1            0               0   
1   700.0  ...         1            0         0            0               0   
2  2990.0  ...         0            0         0            0               0   
3   420.0  ...         0            0         0            0               0   
4   450.0  ...         0            0         0            0               0   

   fuel_가솔린/LPG겸용  fuel_기타  fuel_디젤  fuel_전기      dist  
0               0        0        0        0   40000.0  
1               0        0        0        0   90000.0  
2               0        0        1        0   30000.0  
3               0        0        1        0  160000.0  
4               0        0        1        0  160000.0  

[5 rows x 23 columns]

# 6 인승 people 765 non-null float64

- 결측치 너무 많다 드랍하자

In [61]:
train = train.drop(['people'], axis=1)
test = test.drop(['people'], axis=1)

# 7 최대출력(마력) power 10644 non-null float64

- 결측치만 채워서 그대로 쓰자

In [62]:
train_power = train['power']
test_power = test['power']

train_power

0        201.0
1        157.0
2        202.0
3        151.0
4        103.0
         ...  
11764    210.0
11765      NaN
11766    370.0
11767    425.0
11768    425.0
Name: power, Length: 11769, dtype: float64

In [63]:
train_power.isna().sum()

1125

In [64]:
train_power = train_power.fillna(train_power.mean())
test_power = test_power.fillna(test_power.mean())

In [65]:
train_power.isna().sum()

0

In [66]:
train_power.name = 'power'
test_power.name = 'power'

train = train.drop(['power'], axis=1)
test = test.drop(['power'], axis=1)

train = train.join(train_power)
test = test.join(test_power)

In [67]:
train.head()

yeartype  cylinder  torque type mission country  newprice   price    year  \
0    2013.0       4.0    25.5   FF     NaN      국산    3141.0  1870.0  2012.0   
1    2010.0       NaN    20.0   FF     NaN      국산       NaN   700.0  2010.0   
2    2015.0       4.0    45.0   FF     NaN      국산    3389.0  2990.0  2014.0   
3    2008.0       4.0    33.8   FR     NaN      국산    2499.0   420.0  2007.0   
4    2004.0       NaN    24.0   FR      수동      국산    1445.0   450.0  2004.0   

   month  ...  fuel_LPG+전기  fuel_가솔린  fuel_가솔린+전기  fuel_가솔린/CNG겸용  \
0   12.0  ...            0         1            0               0   
1    3.0  ...            0         0            0               0   
2    9.0  ...            0         0            0               0   
3    6.0  ...            0         0            0               0   
4    3.0  ...            0         0            0               0   

   fuel_가솔린/LPG겸용  fuel_기타  fuel_디젤  fuel_전기      dist  power  
0               0        0        0        0   40000.0  201.0  
1               0        0        0        0   90000.0  157.0  
2               0        0        1        0   30000.0  202.0  
3               0        0        1        0  160000.0  151.0  
4               0        0        1        0  160000.0  103.0  

[5 rows x 22 columns]

# 8 기통 cylinder 8698 non-null float64

- 범주형으로 조질 수 있겠다
- 단 결측치는 어떻게 할 것인가?
- 최빈값으로 일단 조져보자



In [68]:
train_cylinder = train['cylinder']
test_cylinder = test['cylinder']

train_cylinder.value_counts()

4.0    5621
6.0    2459
3.0     402
8.0     171
5.0      45
Name: cylinder, dtype: int64

In [69]:
test_cylinder.value_counts()

4.0    2747
6.0    1222
3.0     173
8.0      95
5.0      38
Name: cylinder, dtype: int64

In [70]:
train_cylinder = train_cylinder.fillna(4.0)
test_cylinder = test_cylinder.fillna(4.0)

In [71]:
train_cylinder.isna().sum()

0

In [72]:
train_cylinder_dummies = pd.get_dummies(train_cylinder)
test_cylinder_dummies = pd.get_dummies(test_cylinder)

# MISSING COLUMN PROCESSING
train_missing_cols = set(test_cylinder_dummies.columns) - set(train_cylinder_dummies.columns)
for c in train_missing_cols:
    train_cylinder_dummies[c] = 0
train_cylinder_dummies = train_cylinder_dummies[test_cylinder_dummies.columns]
test_missing_cols = set(train_cylinder_dummies.columns) - set(test_cylinder_dummies.columns)
for c in test_missing_cols:
    test_cylinder_dummies[c] = 0
test_cylinder_dummies = test_cylinder_dummies[train_cylinder_dummies.columns]
# MISSING COLUMN PROCESSING

train_cylinder_dummies = train_cylinder_dummies.add_prefix("cylinder_")
test_cylinder_dummies = test_cylinder_dummies.add_prefix("cylinder_")

train = train.drop(['cylinder'], axis=1)
test = test.drop(['cylinder'], axis=1)

train = train.join(train_cylinder_dummies)
test = test.join(test_cylinder_dummies)

In [73]:
train.head()

yeartype  torque type mission country  newprice   price    year  month  \
0    2013.0    25.5   FF     NaN      국산    3141.0  1870.0  2012.0   12.0   
1    2010.0    20.0   FF     NaN      국산       NaN   700.0  2010.0    3.0   
2    2015.0    45.0   FF     NaN      국산    3389.0  2990.0  2014.0    9.0   
3    2008.0    33.8   FR     NaN      국산    2499.0   420.0  2007.0    6.0   
4    2004.0    24.0   FR      수동      국산    1445.0   450.0  2004.0    3.0   

   fuel_CNG  ...  fuel_기타  fuel_디젤  fuel_전기      dist  power  cylinder_3.0  \
0         0  ...        0        0        0   40000.0  201.0             0   
1         0  ...        0        0        0   90000.0  157.0             0   
2         0  ...        0        1        0   30000.0  202.0             0   
3         0  ...        0        1        0  160000.0  151.0             0   
4         0  ...        0        1        0  160000.0  103.0             0   

   cylinder_4.0  cylinder_5.0  cylinder_6.0  cylinder_8.0  
0             1             0             0             0  
1             1             0             0             0  
2             1             0             0             0  
3             1             0             0             0  
4             1             0             0             0  

[5 rows x 26 columns]

# 9 최대토크(kgm) torque 10545 non-null float64

- 얘도 결측치 mean으로 채우고 그대로 쓰자

In [74]:
train_torque = train['torque']
test_torque = test['torque']

train_torque

0        25.5
1        20.0
2        45.0
3        33.8
4        24.0
         ... 
11764    30.5
11765     NaN
11766    52.0
11767    53.0
11768    53.0
Name: torque, Length: 11769, dtype: float64

In [75]:
train_torque = train_torque.fillna(train_torque.mean())
test_torque = test_torque.fillna(test_torque.mean())

In [76]:
train_torque.isna().sum()

0

In [77]:
train_torque.name = 'torque'
test_torque.name = 'torque'

train = train.drop(['torque'], axis=1)
test = test.drop(['torque'], axis=1)

train = train.join(train_torque)
test = test.join(test_torque)

In [78]:
train.head()

yeartype type mission country  newprice   price    year  month  fuel_CNG  \
0    2013.0   FF     NaN      국산    3141.0  1870.0  2012.0   12.0         0   
1    2010.0   FF     NaN      국산       NaN   700.0  2010.0    3.0         0   
2    2015.0   FF     NaN      국산    3389.0  2990.0  2014.0    9.0         0   
3    2008.0   FR     NaN      국산    2499.0   420.0  2007.0    6.0         0   
4    2004.0   FR      수동      국산    1445.0   450.0  2004.0    3.0         0   

   fuel_LPG  ...  fuel_디젤  fuel_전기      dist  power  cylinder_3.0  \
0         0  ...        0        0   40000.0  201.0             0   
1         1  ...        0        0   90000.0  157.0             0   
2         0  ...        1        0   30000.0  202.0             0   
3         0  ...        1        0  160000.0  151.0             0   
4         0  ...        1        0  160000.0  103.0             0   

   cylinder_4.0  cylinder_5.0  cylinder_6.0  cylinder_8.0  torque  
0             1             0             0             0    25.5  
1             1             0             0             0    20.0  
2             1             0             0             0    45.0  
3             1             0             0             0    33.8  
4             1             0             0             0    24.0  

[5 rows x 26 columns]

# 10 구동방식 type 10808 non-null object

- 제일 많은 걸로 결측치 채우고 dummies 만들어서 쓰자

In [79]:
train_type = train['type']
test_type = test['type']

train_type.value_counts()

FF     7333
FR     2319
4WD     688
AWD     452
MR       13
RR        3
Name: type, dtype: int64

In [80]:
train_type = train_type.fillna('FF')
test_type = test_type.fillna('FF')

In [81]:
train_type.isna().sum()

0

In [82]:
train_type_dummies = pd.get_dummies(train_type)
test_type_dummies = pd.get_dummies(test_type)

# MISSING COLUMN PROCESSING
train_missing_cols = set(test_type_dummies.columns) - set(train_type_dummies.columns)
for c in train_missing_cols:
    train_type_dummies[c] = 0
train_type_dummies = train_type_dummies[test_type_dummies.columns]
test_missing_cols = set(train_type_dummies.columns) - set(test_type_dummies.columns)
for c in test_missing_cols:
    test_type_dummies[c] = 0
test_type_dummies = test_type_dummies[train_type_dummies.columns]
# MISSING COLUMN PROCESSING

train_type_dummies = train_type_dummies.add_prefix('type_')
test_type_dummies = test_type_dummies.add_prefix('type_')

train = train.drop(['type'], axis=1)
test = test.drop(['type'], axis=1)

train = train.join(train_type_dummies)
test = test.join(test_type_dummies)

In [83]:
train.head()

yeartype mission country  newprice   price    year  month  fuel_CNG  \
0    2013.0     NaN      국산    3141.0  1870.0  2012.0   12.0         0   
1    2010.0     NaN      국산       NaN   700.0  2010.0    3.0         0   
2    2015.0     NaN      국산    3389.0  2990.0  2014.0    9.0         0   
3    2008.0     NaN      국산    2499.0   420.0  2007.0    6.0         0   
4    2004.0      수동      국산    1445.0   450.0  2004.0    3.0         0   

   fuel_LPG  fuel_LPG+전기  ...  cylinder_4.0  cylinder_5.0  cylinder_6.0  \
0         0            0  ...             1             0             0   
1         1            0  ...             1             0             0   
2         0            0  ...             1             0             0   
3         0            0  ...             1             0             0   
4         0            0  ...             1             0             0   

   cylinder_8.0  torque  type_4WD  type_AWD  type_FF  type_FR  type_MR  
0             0    25.5         0         0        1        0        0  
1             0    20.0         0         0        1        0        0  
2             0    45.0         0         0        1        0        0  
3             0    33.8         0         0        0        1        0  
4             0    24.0         0         0        0        1        0  

[5 rows x 30 columns]

# 11 자동수동 mission 1313 non-null object

- 못 쓴다 드랍하자

In [84]:
train = train.drop(['mission'], axis=1)
test = test.drop(['mission'], axis=1)

In [85]:
train.head()

yeartype country  newprice   price    year  month  fuel_CNG  fuel_LPG  \
0    2013.0      국산    3141.0  1870.0  2012.0   12.0         0         0   
1    2010.0      국산       NaN   700.0  2010.0    3.0         0         1   
2    2015.0      국산    3389.0  2990.0  2014.0    9.0         0         0   
3    2008.0      국산    2499.0   420.0  2007.0    6.0         0         0   
4    2004.0      국산    1445.0   450.0  2004.0    3.0         0         0   

   fuel_LPG+전기  fuel_가솔린  ...  cylinder_4.0  cylinder_5.0  cylinder_6.0  \
0            0         1  ...             1             0             0   
1            0         0  ...             1             0             0   
2            0         0  ...             1             0             0   
3            0         0  ...             1             0             0   
4            0         0  ...             1             0             0   

   cylinder_8.0  torque  type_4WD  type_AWD  type_FF  type_FR  type_MR  
0             0    25.5         0         0        1        0        0  
1             0    20.0         0         0        1        0        0  
2             0    45.0         0         0        1        0        0  
3             0    33.8         0         0        0        1        0  
4             0    24.0         0         0        0        1        0  

[5 rows x 29 columns]

# 12 국산/수입 country 11769 non-null object

- 단순한 범주형 같다
- 아니 근데 train이랑 test 다 국산밖에 없는데?
- 드랍하자...

In [86]:
train_country = train['country']
test_country = test['country']

train_country.value_counts()

국산    11769
Name: country, dtype: int64

In [87]:
test_country.value_counts()

국산    5789
Name: country, dtype: int64

In [88]:
train = train.drop(['country'], axis=1)
test = test.drop(['country'], axis=1)

In [89]:
train.head()

yeartype  newprice   price    year  month  fuel_CNG  fuel_LPG  fuel_LPG+전기  \
0    2013.0    3141.0  1870.0  2012.0   12.0         0         0            0   
1    2010.0       NaN   700.0  2010.0    3.0         0         1            0   
2    2015.0    3389.0  2990.0  2014.0    9.0         0         0            0   
3    2008.0    2499.0   420.0  2007.0    6.0         0         0            0   
4    2004.0    1445.0   450.0  2004.0    3.0         0         0            0   

   fuel_가솔린  fuel_가솔린+전기  ...  cylinder_4.0  cylinder_5.0  cylinder_6.0  \
0         1            0  ...             1             0             0   
1         0            0  ...             1             0             0   
2         0            0  ...             1             0             0   
3         0            0  ...             1             0             0   
4         0            0  ...             1             0             0   

   cylinder_8.0  torque  type_4WD  type_AWD  type_FF  type_FR  type_MR  
0             0    25.5         0         0        1        0        0  
1             0    20.0         0         0        1        0        0  
2             0    45.0         0         0        1        0        0  
3             0    33.8         0         0        0        1        0  
4             0    24.0         0         0        0        1        0  

[5 rows x 28 columns]

# 13 신차가(만원) newprice 9544 non-null float64

- 일단 드랍


In [90]:
train = train.drop(['newprice'], axis=1)

In [91]:
train.head()

yeartype   price    year  month  fuel_CNG  fuel_LPG  fuel_LPG+전기  fuel_가솔린  \
0    2013.0  1870.0  2012.0   12.0         0         0            0         1   
1    2010.0   700.0  2010.0    3.0         0         1            0         0   
2    2015.0  2990.0  2014.0    9.0         0         0            0         0   
3    2008.0   420.0  2007.0    6.0         0         0            0         0   
4    2004.0   450.0  2004.0    3.0         0         0            0         0   

   fuel_가솔린+전기  fuel_가솔린/CNG겸용  ...  cylinder_4.0  cylinder_5.0  cylinder_6.0  \
0            0               0  ...             1             0             0   
1            0               0  ...             1             0             0   
2            0               0  ...             1             0             0   
3            0               0  ...             1             0             0   
4            0               0  ...             1             0             0   

   cylinder_8.0  torque  type_4WD  type_AWD  type_FF  type_FR  type_MR  
0             0    25.5         0         0        1        0        0  
1             0    20.0         0         0        1        0        0  
2             0    45.0         0         0        1        0        0  
3             0    33.8         0         0        0        1        0  
4             0    24.0         0         0        0        1        0  

[5 rows x 27 columns]

# 14 가격(만원) price 11769 non-null float64

- 이건 피쳐가 아니고 정답(레이블)

# 종합

In [92]:
print(train.shape)
print(test.shape)

(11769, 27)
(5789, 26)


# 데이터 나누기

In [93]:
X_train = train.drop('price', axis=1)
Y_train = train['price']
X_test = test.copy()

print(X_train.shape)
print(Y_train.shape)
print(X_test.shape)

(11769, 26)
(11769,)
(5789, 26)


In [94]:
from sklearn.linear_model import ElasticNet, Lasso,  BayesianRidge, LassoLarsIC
from sklearn.ensemble import RandomForestRegressor,  GradientBoostingRegressor
from sklearn.kernel_ridge import KernelRidge
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import RobustScaler
from sklearn.base import BaseEstimator, TransformerMixin, RegressorMixin, clone
from sklearn.model_selection import KFold, cross_val_score, train_test_split
from sklearn.metrics import mean_squared_error
from xgboost import XGBRegressor
import xgboost as xgb
import lightgbm as lgb
from sklearn.svm import SVR
from sklearn.ensemble import RandomForestRegressor
from sklearn.datasets import make_regression
from sklearn.model_selection import GridSearchCV

In [95]:
n_folds = 10

def rmsle_cv(model):
    kf = KFold(n_folds, shuffle=True, random_state=42).get_n_splits(X_train.values)
    rmse= np.sqrt(-cross_val_score(model, X_train.values, Y_train.values, scoring="neg_mean_squared_error", cv = kf))
    return(rmse)

In [96]:
lasso = make_pipeline(RobustScaler(), Lasso(alpha =0.0005, random_state=3))

In [97]:
ENet = make_pipeline(RobustScaler(), ElasticNet(alpha=0.0005, l1_ratio=.9, random_state=3))


In [98]:
KRR = KernelRidge(alpha=0.6, kernel='polynomial', degree=2, coef0=2.5) #kernel = 'rbf' , 'sigmoid' 

In [99]:
GBoost = GradientBoostingRegressor(n_estimators=3000, learning_rate=0.05,
                                   max_depth=4, max_features='sqrt',
                                   min_samples_leaf=15, min_samples_split=10, 
                                   loss='huber', random_state =5)

In [100]:
model_xgb = xgb.XGBRegressor(colsample_bytree=0.4603, gamma=0.0468, 
                             learning_rate=0.05, max_depth=3, 
                             min_child_weight=1.7817, n_estimators=2200,
                             reg_alpha=0.4640, reg_lambda=0.8571,
                             subsample=0.5213, silent=1,
                             random_state =7, nthread = -1)

In [101]:
model_lgb = lgb.LGBMRegressor(objective='regression',num_leaves=5,
                              learning_rate=0.05, n_estimators=720,
                              max_bin = 55, bagging_fraction = 0.8,
                              bagging_freq = 5, feature_fraction = 0.2319,
                              feature_fraction_seed=9, bagging_seed=9,
                              min_data_in_leaf =6, min_sum_hessian_in_leaf = 11)

In [102]:
model_svr = SVR(C=1, cache_size=200, coef0=0, degree=3, epsilon=0.0, gamma='auto', kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

In [103]:
params = {'coef0':[0, 0.1, 0.5, 1], 'C' :[0.1, 0.2, 0.5, 1], 'epsilon':[i/10.0 for i in range(0,6)]}

In [104]:
regr = RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features=150, max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=90, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [105]:
param_grid = [
    {'n_estimators': [3, 10, 30, 60, 90], 'max_features': [50,100,150,200,250,300]},
    {'bootstrap': [True], 'n_estimators': [3, 10, 30, 60, 90], 'max_features': [50,100,150,200,250]},
]

In [106]:
score = rmsle_cv(lasso)
print("\nLasso score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))


Lasso score: 1131.6794 (844.3153)



In [107]:
score = rmsle_cv(ENet)
print("ElasticNet score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

ElasticNet score: 1131.3250 (844.4425)



In [96]:
# score = rmsle_cv(KRR)
# print("Kernel Ridge score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

In [108]:
score = rmsle_cv(GBoost)
print("Gradient Boosting score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

Gradient Boosting score: 929.1261 (886.4515)



In [109]:
score = rmsle_cv(model_xgb)
print("Xgboost score: {:.4f} ({:.4f})\n".format(score.mean(), score.std()))

Xgboost score: 999.5109 (861.4094)



In [110]:
# score = rmsle_cv(model_lgb)
# print("LGBM score: {:.4f} ({:.4f})\n" .format(score.mean(), score.std()))

In [111]:
score = rmsle_cv(model_svr)
print("SVR score: {:.4f} ({:.4f})\n" .format(score.mean(), score.std()))

SVR score: 1626.0520 (779.9771)



In [112]:
score = rmsle_cv(regr)
print("RF score: {:.4f} ({:.4f})\n" .format(score.mean(), score.std()))

RF score: nan (nan)



# Prediction

In [113]:
def rmsle(Y, Y_pred): 
    return np.sqrt(mean_squared_error(Y, Y_pred))

In [119]:
model_xgb.fit(X_train, Y_train)
xgb_train_pred = model_xgb.predict(X_train)
# xgb_pred = np.expm1(model_xgb.predict(X_test))
xgb_pred = model_xgb.predict(X_test)
print(rmsle(Y_train, xgb_train_pred))

1191.839408164097


In [117]:
# model_lgb.fit(X_train, Y_train)
# lgb_train_pred = model_lgb.predict(X_train)
# lgb_pred = np.expm1(model_lgb.predict(X_test.values))
# print(rmsle(Y_train, lgb_train_pred))

In [121]:
GBoost.fit(X_train, Y_train)
GB_train_pred = GBoost.predict(X_train)
# GB_pred = np.expm1(GBoost.predict(X_test.values))
GB_pred = GBoost.predict(X_test)
print(rmsle(Y_train, GB_train_pred))

1251.2084114348754


In [123]:
model_svr.fit(X_train, Y_train)
svr_train_pred = model_svr.predict(X_train)
svr_pred = model_svr.predict(X_test)
print(rmsle(Y_train, svr_train_pred))

1802.0575946068993


# Ensemble

In [122]:
GB_pred

array([5455.40450124, 1116.23429809, 1212.28476405, ...,  743.73259282,
       1233.57607498, 2737.94792082])

In [124]:
ensemble = xgb_pred*0.25 + GB_pred*0.25 + svr_pred*0.5

In [125]:
ensemble

array([3758.1394174 , 1613.24957657, 1628.49052558, ..., 1314.04443834,
       1433.68041875, 2295.95434945])

In [126]:
submission = pd.DataFrame({
    "no": test_no,
    "가격(만원)": ensemble
})
submission.to_csv('../output/ensemble.csv', index=False)